### ISSUE TITLE
Async LocalFileSystem

### ISSUE THREAD

**Issue by tbepler** [fsspec/filesystem_spec#1052]
It would be nice to have an async version of LocalFileSystem for compatibility with async remote file systems and reading/writing large files locally without blocking.

Are there any plans to support async for LocalFileSystem or is there some other implementation of this I'm not aware of?

**Comment by martindurant** [comment]
I am not entirely opposed, and there are other issues around talking about this. However, I am not convinced that python async implementations do what you think, since IO is offloaded to OS buffers very quickly - this is not the high-latency situation faced by remote storage. `aiofiles`, I believe, uses threads to appear as though it is running concurrently, but in fact it is not. [aiofile](https://github.com/mosquito/aiofile) claims to be truly asynchronous on linux only.

**Comment by skshetry** [comment]
We have a version implemented in [`morefs`](https://github.com/iterative/morefs): [`AsyncLocalFileSystem`](https://github.com/iterative/morefs/blob/622c87482e1f64fe8c2e8e87218b752e3d851c99/src/morefs/asyn_local.py).

So far we have only used that for testing where the other filesystems were async, but not in real world.
The performance is not great compared to LocalFileSystem, since those are run in the thread by `aiofile`/`aiopath`/`aiofiles`.
```console
pip install morefs[asynclocalfs]
```
```python
from fsspec import filesystem

fs = filesystem("asynclocalfs")
```

**Comment by martindurant** [comment]
I would be happy to have asynclocalfs in the known implementations of fsspec, with these caveats. It would be accessed as "asynclocalfs:///path/file". I would not recommend anyone having it take over the default protocol (i.e., when there is no protocol at all).

**Comment by martindurant** [comment]
I meant keeping it as a separate repo, and amending the registry.

**Comment by skshetry** [comment]
Thanks. I'll add that to the registry then. :)

(I do need a way to not generate sync methods from async APIs so that sync methods fall back to `LocalFileSystem` instead of wrapped ones, I have been hacking it with `async_impl=False`. will propose something.).

**Comment by skshetry** [comment]
With `fsspec==2022.10.0` released, I have also released `morefs==0.1.0` that has `AsyncLocalFileSystem` with some performance improvements and compatibility improvements. Synchronous methods fallback to fsspec's `LocalFileSystem` transparently and async methods just wrap over synchronous methods.

One can use it as follows:
```python
from fsspec import filesystem

fs = filesystem("asynclocal")
```

Or, directly:
```python
from morefs.asyn_local import AsyncLocalFileSystem

fs = AsyncLocalFileSystem(auto_mkdir=...)
```

**Comment by skshetry** [comment]
With `fsspec==2022.10.0` released, I have also released `morefs==0.1.0` that has `AsyncLocalFileSystem` with some performance improvements and compatibility improvements. Synchronous methods fallback to fsspec's `LocalFileSystem` transparently and async methods just wrap over synchronous methods.

One can use it as follows:
```python
from fsspec import filesystem

fs = filesystem("asynclocal")
```

Or, directly:
```python
from morefs.asyn_local import AsyncLocalFileSystem

fs = AsyncLocalFileSystem(auto_mkdir=...)
```
